### PR TITLE
TDI-37601: Can't find the option 'create or update' on tDatasetOutput

### PR DESCRIPTION
**Pull request by wwang-talend**
https://jira.talendforge.org/browse/TDI-37601

### CODE REVIEW

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/213/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/213/)

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java [hunk 1]
```diff
@@ -256,4 +258,22 @@ private boolean isRequiredFieldRight() {
...
+        return 1;
+    }
+
+    @Override
+    public boolean postDeserialize(int version, PostDeserializeSetup setup, boolean persistent) {
+        boolean migrated = super.postDeserialize(version, setup, persistent);
+
+        if (version < this.getVersionNumber()) {
+            mode.setPossibleValues(DataPrepOutputModes.class.getEnumConstants());
+            migrated = true;
```
Comment (gonchar-ivan) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: Consider to change it:
migrated = migrated && true;   // to take into account result of super class migration

Comment (wwang-talend) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: should use "or", not "and", so this is right.

Comment (gonchar-ivan) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: Sorry, I was wrong.
postDeserialize() returns true if object was changed during postDeserialize()
(I thought this returns something like "correctly deserialized". So false value from super class means fail of deserialization)
So, above code is right

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java [hunk 2]
```diff
@@ -256,4 +258,22 @@ private boolean isRequiredFieldRight() {
...
                 && !isEmpty(dataSetName.getStringValue()) && !isEmpty(dataSetId.getStringValue());
     }
+
+    @Override
+    public int getVersionNumber() {
+        return 1;
+    }
+
+    @Override
+    public boolean postDeserialize(int version, PostDeserializeSetup setup, boolean persistent) {
```
Comment (gonchar-ivan) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: Possible way is to implement private/protected migrateToX() for each new version.
migrateToX() should call migrateTo(X-1) in the 1st line. (something like recursion)
Then in postDeserialize() we can have 
switch(version) {
case X-1: migrateTo(X-1);
case X: migrateToX;
}

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/235/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/235/)

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java (hunk 2 above)
Comment (wwang-talend) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: agree totally.  Another thing :  now we don't write the version information to the serialized json as some bug exists, please see the junit case in the dev branch :  https://github.com/Talend/daikon/commit/90a8ffedca119a3e9e18c9411ea8c1f4530bd68a

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java [hunk 3]
```diff
@@ -256,4 +258,22 @@ private boolean isRequiredFieldRight() {
         return !isEmpty(url.getStringValue()) && !isEmpty(login.getStringValue()) && !isEmpty(pass.getStringValue())
                 && !isEmpty(dataSetName.getStringValue()) && !isEmpty(dataSetId.getStringValue());
     }
+
+    @Override
+    public int getVersionNumber() {
```
Comment (wwang-talend) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: yes, only follow the usage in the junit test. In fact, i want to make the version setting works, but fails.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/243/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/243/)

**Comment by wwang-talend** [comment]
have added the junit test case

**Comment by wwang-talend** [comment]
have rebase it to master

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/261/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/261/)

**Comment by gonchar-ivan** [comment]
PR was fixed according Seb's review. I also look through it. Looks good to me. So, I'll merge.

**Comment by sgandon** [comment]
Please do not merge this as I suggested another fix in daikon : https://github.com/Talend/daikon/commit/28853df1861d2fafcb6e40bd8739a817899e46bf
So close this if the fix in daikon is checked.

**Comment by wwang-talend** [comment]
Hi Seb, i try your fix in the studio, can't work, have shown you in the skype. And as the code freeze will appear in one hour, i will merge this commit now.

**Comment by pteyssier** [comment]
Need to revert this code after 6.3.0 Beta and Daikon issue fixed.

**Comment by sgandon** [comment]
ok

**Pull request merged**